### PR TITLE
[ASV-853] New download manager core implementation

### PR DESCRIPTION
**Pull request by phillipsgoncalves**
**What does this PR do?**

This implements the new download manager core [Downloads Refactor Presentation link](https://docs.google.com/presentation/d/1noLMtJQXcMba8pcsHNXbxHDZg-1J7FMgPKB_AU0DxV8/edit?usp=sharing) .
This core is represented by the DownloadManager (manages the downloads as whole), AppDownloader (manages App downloads) and FileDownloader (manages the download of a file and communicates with the lib).
Also, created two tests for the AppDownloader layer and another one for the FileDownloader layer.

**Database changed?**

   Yes | No

**Where should the reviewer start?**

- [ ] NewDownloadManager.java
- [ ] AppDownloadManager.java
- [ ] FileDownloadManager.java

**How should this be manually tested?**

This is not visible yet, so the tests should be regarding architecture.

**What are the relevant tickets?**

  Tickets related to this pull-request: [ASV-853](https://aptoide.atlassian.net/browse/ASV-853)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new libs, new keys, etc.) 

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Partners build
- [ ] Unit tests pass
- [ ] Functional QA tests pass

### CODE REVIEW

File: downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/NewAptoideDownloadManager.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+/**
+ * Created by filipegoncalves on 7/27/18.
+ */
+
+public class NewAptoideDownloadManager implements DownloadManager {
+
+  public static final int PROGRESS_MAX_VALUE = 100;
```
Comment (jdandrade) [inline downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/NewAptoideDownloadManager.java]: duplicated constant - not being used.

Comment (phillipsgoncalves) [reply downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/NewAptoideDownloadManager.java]: done

File: downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/FileDownloadCallback.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package cm.aptoide.pt.downloadmanager;
+
+/**
+ * Created by filipegoncalves on 8/29/18.
+ */
+
+public interface FileDownloadCallback {
+  int getDownloadProgress();
+
+  int getType();
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/FileDownloadCallback.java]: done

File: downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/DownloadStatusMapper.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        break;
+      case PAUSED:
+        downloadState = Download.PAUSED;
+        break;
+      case WARN:
+        downloadState = Download.WARN;
+        break;
+      case ERROR:
+        downloadState = Download.ERROR;
+        break;
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/DownloadStatusMapper.java]: done

File: downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/AppDownloadStatus.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    return appDownloadState;
+  }
+
+  public int getDownloadSpeed() {
+    return downloadSpeed;
+  }
+
+  public void setDownloadSpeed(int downloadSpeed) {
+    this.downloadSpeed = downloadSpeed;
+  }
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/AppDownloadStatus.java]: done

File: downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/AppDownloadStatus.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  private int downloadSpeed;
+  private AppDownloadState appDownloadState;
+
+  public AppDownloadStatus(String md5, FileDownloadCallback apk, FileDownloadCallback obbMain,
+      FileDownloadCallback obbPatch, int downloadSpeed, AppDownloadState appDownloadState) {
+    this.md5 = md5;
+    this.apk = apk;
+    this.obbMain = obbMain;
+    this.obbPatch = obbPatch;
+    this.downloadSpeed = downloadSpeed;
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply downloadmanager/src/main/java/cm/aptoide/pt/downloadmanager/AppDownloadStatus.java]: removed it. Soon i will remove it from the database too.

File: app/src/main/java/cm/aptoide/pt/download/DownloadTaskStatus.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * Created by filipegoncalves on 8/29/18.
+ */
+
+public class DownloadTaskStatus implements FileDownloadCallback {
+
+  private AppDownloadState appDownloadState;
+  private int downloadProgress;
+  private int fileType;
+
+  public DownloadTaskStatus(AppDownloadState state, int downloadProgress, int fileType) {
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply app/src/main/java/cm/aptoide/pt/download/DownloadTaskStatus.java]: done

File: app/src/main/java/cm/aptoide/pt/download/DownloadTaskStatus.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package cm.aptoide.pt.download;
+
+import cm.aptoide.pt.downloadmanager.AppDownloadStatus.AppDownloadState;
+import cm.aptoide.pt.downloadmanager.FileDownloadCallback;
+
+/**
+ * Created by filipegoncalves on 8/29/18.
+ */
+
+public class DownloadTaskStatus implements FileDownloadCallback {
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply app/src/main/java/cm/aptoide/pt/download/DownloadTaskStatus.java]: done

**Pull request merged**